From: Steve Best <sbest@redhat.com>
Date: Mon, 2 Aug 2010 20:47:45 -0400
Subject: [ppc] vio: add power management support
Message-id: <20100802203636.18096.17325.sendpatchset@squad5-lp1.lab.bos.redhat.com>
Patchwork-id: 27289
O-Subject: [PATCH RHEL5.6 BZ565570 2/9] powerpc/vio: Add power management support
Bugzilla: 565570
RH-Acked-by: David Howells <dhowells@redhat.com>

RHBZ#:
======
https://bugzilla.redhat.com/show_bug.cgi?id=565570

Description:
============
Adds support for suspend/resume for VIO devices. This is needed for
support for Hardware Management Console (HMC) initiated hibernation.

Signed-off-by: Brian King <brking@linux.vnet.ibm.com>
Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>

RHEL Version Found:
===================
5.6

kABI Status:
============
No symbols were harmed.

Brew:
=====
Built on all platforms. All patches brewed together.
http://brewweb.devel.redhat.com/brew/taskinfo?taskID=2642470

Upstream Status:
================
commit id:  http://git.kernel.org/gitweb.cgi?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=a32aaf14513da776556ad9995de8d83cd76ae60a

===============================================================
Steve Best
IBM on-site partner

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/vio.c b/arch/powerpc/kernel/vio.c
index 0b2b07c..4a1c83e 100644
--- a/arch/powerpc/kernel/vio.c
+++ b/arch/powerpc/kernel/vio.c
@@ -444,6 +444,22 @@ static int vio_hotplug(struct device *dev, char **envp, int num_envp,
 	return 0;
 }
 
+static int vio_bus_suspend(struct device *dev, pm_message_t state)
+{
+	if (dev->driver && dev->driver->suspend)
+		return dev->driver->suspend(dev, state);
+
+	return 0;
+}
+
+static int vio_bus_resume(struct device *dev)
+{
+	if (dev->driver && dev->driver->resume)
+		return dev->driver->resume(dev);
+
+	return 0;
+}
+
 struct bus_type vio_bus_type = {
 	.name = "vio",
 	.dev_attrs = vio_dev_attrs,
@@ -452,6 +468,8 @@ struct bus_type vio_bus_type = {
 	.probe = vio_bus_probe,
 	.remove = vio_bus_remove,
 	.shutdown = vio_bus_shutdown,
+	.suspend = vio_bus_suspend,
+	.resume = vio_bus_resume,
 };
 
 /**
